opal/mca/btl/tcp/btl_tcp_endpoint.c

```diff
@@ -573,44 +582,51 @@ static void mca_btl_tcp_endpoint_connected(mca_btl_base_endpoint_t* btl_endpoint
  */
 static int mca_btl_tcp_endpoint_recv_blocking(mca_btl_base_endpoint_t* btl_endpoint, void* data, size_t size)
 {
-    unsigned char* ptr = (unsigned char*)data;
-    size_t cnt = 0;
-    while(cnt < size) {
-        int retval = recv(btl_endpoint->endpoint_sd, (char *)ptr+cnt, size-cnt, 0);
-
-        /* remote closed connection */
-        if(retval == 0) {
-            mca_btl_tcp_endpoint_close(btl_endpoint);
-            return cnt;
-        }
-
-        /* socket is non-blocking so handle errors */
-        if(retval < 0) {
-            if(opal_socket_errno != EINTR && opal_socket_errno != EAGAIN && opal_socket_errno != EWOULDBLOCK) {
-                BTL_ERROR(("recv(%d, %lu/%lu) failed: %s (%d)",
-                           btl_endpoint->endpoint_sd, cnt, size, strerror(opal_socket_errno), opal_socket_errno));
-                btl_endpoint->endpoint_state = MCA_BTL_TCP_FAILED;
-                mca_btl_tcp_endpoint_close(btl_endpoint);
-                return -1;
-            }
-            continue;
-        }
-        cnt += retval;
+    int ret = mca_btl_tcp_recv_blocking(btl_endpoint->endpoint_sd, data, size);
+    if (ret <= 0) {
+        mca_btl_tcp_endpoint_close(btl_endpoint);
     }
-    return cnt;
+    return ret;
 }
 
 
 /*
  *  Receive the endpoints globally unique process identification from a newly
  *  connected socket and verify the expected response. If so, move the
  *  socket to a connected state.
+ *
+ *  NOTE: The return codes from this function are checked in
+ *  mca_btl_tcp_endpoint_recv_handler().  Don't change them here
+ *  without also changing the handling in _recv_handler()!
  */
 static int mca_btl_tcp_endpoint_recv_connect_ack(mca_btl_base_endpoint_t* btl_endpoint)
 {
-    size_t s;
+    size_t s, len = strlen(mca_btl_tcp_magic_id_string);;
     opal_process_name_t guid;
     mca_btl_tcp_proc_t* btl_proc = btl_endpoint->endpoint_proc;
+    char msg[1024];
+
+    /* First get magic string indicating that the connector is an Open MPI TCP BTL */
+    assert(len < sizeof(msg));
+    msg[0] = '\0';
+    s = mca_btl_tcp_endpoint_recv_blocking(btl_endpoint, msg, len);
```

Comment:
Similar to the send side, this 2 steps operation can be done with a single access to the socket.